src/coreclr/src/jit/optimizer.cpp

```diff
@@ -8659,6 +8659,163 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
     return true;
 }
 
+//----------------------------------------------------------------------------------
+// optBranchlessConditions: optimize simple expressions to branchless expressions
+//
+//  Operation:
+//      Look for BBJ_COND basic blocks which target two BBJ_RETURN blocks
+//      and convert them into single branchless expressions, e.g.
+//
+//            JTRUE
+//              \
+//              EQ              RETURN
+//             /  \                \
+//            X    Y               ADD
+//                       =>       /   \
+//           RETURN             EQ     C3
+//              \              /  \
+//              C1            X    Y
+//
+//           RETURN
+//              \
+//               C2
+//
+//  Note:
+//      At the moment, the optimization works only for simple BBJ_RETURN blocks
+//      and returned values are of difference of 1.
+//
+void Compiler::optBranchlessConditions()
+{
+    JITDUMP("\n*************** In optBranchlessConditions()\n");
+    bool changed = false;
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+    {
+        if ((block->bbJumpKind != BBJ_COND))
+        {
+            continue;
+        }
+
+        Statement* firstStmt = block->lastStmt();
+        if ((firstStmt == nullptr) || !firstStmt->GetRootNode()->OperIs(GT_JTRUE))
+        {
+            continue;
+        }
+
+        GenTree*  rootNode    = firstStmt->GetRootNode();
+        GenTree*  rootSubNode = rootNode->gtGetOp1();
+        var_types typ         = rootSubNode->TypeGet();
+
+        assert(rootSubNode != nullptr);
+
+        if (!varTypeIsIntegral(typ) || (rootNode->gtFlags & GTF_ORDER_SIDEEFF))
+        {
+            continue;
+        }
+
+        BasicBlock* trueBb  = block->bbJumpDest;
+        BasicBlock* falseBb = block->bbNext;
+
+        assert((trueBb != nullptr) && (falseBb != nullptr));
+
+        if (!rootSubNode->OperIsCompare() || (trueBb == falseBb) ||
+            // both blocks are expected to be BBJ_RETURN
+            // TODO: implement for GT_ASG
+            (trueBb->bbJumpKind != BBJ_RETURN) || (falseBb->bbJumpKind != BBJ_RETURN) ||
+            // give up if these blocks are used by someone else
+            (trueBb->countOfInEdges() != 1) || (falseBb->countOfInEdges() != 1) ||
```

Comment:
> // give up if these blocks are used by someone else

why?

Reply:
Need to check but I think there will be regressions (in terms of size, not performance)

Reply:
The transformation will be invalid, because those return blocks can be reached under other conditions.

Reply:
> The transformation will be invalid

Why? We don't delete these blocks, we only remove the edges from `block` to `trueBb` and `falseBb`. `trueBb` and `falseBb` are staying alive for other preds.